apps/internal/base/base.go

```diff
@@ -302,3 +306,30 @@ func convertStrUnixToUTCTime(unixTime string) (time.Time, error) {
 	}
 	return time.Unix(timeInt, 0).UTC(), nil
 }
+
+func getKeyFromRequest(authParams authority.AuthParams) string {
+	if authParams.AuthorizationType == authority.ATClientCredentials {
+		return getAppKey(authParams)
+	}
+	if authParams.AuthorizationType == authority.ATRefreshToken {
+		return authParams.HomeaccountID
+	}
+	return ""
+}
+
+func getKeyFromResponse(authParams authority.AuthParams, homeaccountIDfromResponse string) string {
+	if authParams.AuthorizationType == authority.ATClientCredentials {
+		return getAppKey(authParams)
+	}
+	if authParams.IsConfidentialClient || authParams.AuthorizationType == authority.ATRefreshToken {
+		return homeaccountIDfromResponse
+	}
+	return ""
+}
+
+func getAppKey(authParams authority.AuthParams) string {
+	if authParams.AuthorityInfo.Tenant != "" {
+		return fmt.Sprintf("%s_%s_AppTokenCache", authParams.ClientID, authParams.AuthorityInfo.Tenant)
+	}
+	return fmt.Sprintf("%s_%s_AppTokenCache", authParams.ClientID, "")
```

Comment:
why sprintf on the second param if it is always empty?

Reply:
resolved in https://github.com/AzureAD/microsoft-authentication-library-for-go/pull/184/commits/71dea81ddef28df17b4741bef3e8643d75e7ccd9